### PR TITLE
Save ipmi sol output as log file

### PR DESCRIPTION
**Pull request by waynechen55**
* **IPMI** SOL console output records important information about system under test, including boot-up process, error messages, unexpected abnormal output and etc, detailed texts of which likely might not be recorded by screenshot or even video clip. 

* **By** saving full output of ipmi sol console to a log file, which will be uploaded at the end of test run, investigation and examination work can be executed much more easily and conveniently without losing any information.

* **Introduce** test suite setting ```IPMISOL_LOG``` to determine whether ipmi sol ouptut will be recorded into log file ```ipmisol-log```.

* **Add** entry for setting ```IPMISOL_LOG``` in ```doc/backend_vars.asciidoc```.

* **This** pull request should be merged together with https://github.com/os-autoinst/openQA/pull/5699. 

* **Verification Runs:**
  * [**test run without IPMISOL_LOG and ipmisol-log** link1](http://10.67.18.153/tests/2842) [link2](http://10.100.228.134/tests/2842) [link3](http://10.100.204.18/tests/2842) [link4](http://10.149.193.170/tests/2842)
  * [**test run with IPMISOL_LOG=0 and without ipmisol-log** link1](http://10.67.18.153/tests/2845) [link2](http://10.100.228.134/tests/2845) [link3](http://10.100.204.18/tests/2845) [link4](http://10.149.193.170/tests/2845)
  * [**test run with IPMISOL_LOG=1 passed and ipmisol-log** link1](http://10.67.18.153/tests/2834) [link2](http://10.100.228.134/tests/2834) [link3](http://10.100.204.18/tests/2834) [link4](http://10.149.193.170/tests/2834)
  * [**test run with IPMISOL_LOG=1 cancelled and ipmisol-log** link1](http://10.67.18.153/tests/2835) [link2](http://10.100.228.134/tests/2835) [link3](http://10.100.204.18/tests/2835) [link4](http://10.149.193.170/tests/2835)
  * [**test run with IPMISOL_LOG=1 failed and ipmisol-log** link1](http://10.67.18.153/tests/2843) [link2](http://10.100.228.134/tests/2843) [link3](http://10.100.204.18/tests/2843) [link4](http://10.149.193.170/tests/2843)
  * [**test run with IPMISOL_LOG=1 timed-out and ipmisol-log** link1](http://10.67.18.153/tests/2844) [link2](http://10.100.228.134/tests/2844) [link3](http://10.100.204.18/tests/2844) [link4](http://10.149.193.170/tests/2844)
  * [**ipmisol-log records full installation and boot-up process** link1](http://10.67.18.153/tests/2834) [link2](http://10.100.228.134/tests/2834) [link3](http://10.100.204.18/tests/2834) [link4](http://10.149.193.170/tests/2834)
  * [**ipmisol-log records full call trace** link1](http://10.67.18.153/tests/2847) [link2](http://10.100.228.134/tests/2847) [link3](http://10.100.204.18/tests/2847) [link4](http://10.149.193.170/tests/2847)

### CODE REVIEW

**Comment by waynechen55** [comment]
@alice-suse @guoxuguang @Julie-CAO @RoyCai7 @nanzhg @tbaev Please have a look.

**Review by Martchus — COMMENTED** [review]
I assume you have tested this manually. It would make sense to document the new variable.

**Comment by waynechen55** [comment]
> I assume you have tested this manually. It would make sense to document the new variable.

Add entry for setting IPMISOL_LOG in doc/backend_vars.asciidoc.

File: doc/backend_vars.asciidoc [hunk 1]
```diff
@@ -94,6 +94,7 @@ IPMI_MC_RESET_PING_COUNT;integer;1;Ping counts that must be successful after mc
 IPMI_MC_RESET_IPMI_TRIES;integer;3;Maximum number of IPMI command tries that are conducted after mc reset
 IPMI_SOL_PERSISTENT_CONSOLE;boolean;1;Make SOL console persistent and don't reset it, enabled by default
 IPMI_SOL_MAX_RECONNECTS;integer;5;Maximum number of SOL reconnects on connection failure
+IPMISOL_LOG;boolean;1;Record IPMI SOL output into ipmisol-log in consoles/sshXtermIPMI.pm, disabled by default
```
Comment (Martchus) [inline doc/backend_vars.asciidoc:97]: For the sake of consistency:
```suggestion
IPMI_SOL_LOG;boolean;1;Record IPMI SOL output into ipmisol-log in consoles/sshXtermIPMI.pm, disabled by default
```

File: consoles/sshXtermIPMI.pm [hunk 2]
```diff
@@ -17,6 +17,7 @@ sub start_sol ($self) {
     my @command = $self->backend->ipmi_cmdline;
     push(@command, qw(sol activate));
     my $serial = $self->{args}->{serial};
+    push(@command, qw(2>&1 | tee -a ipmisol-log)) if $bmwqemu::vars{IPMISOL_LOG};
```
Comment (Martchus) [inline consoles/sshXtermIPMI.pm:20]:
```suggestion
    push(@command, qw(2>&1 | tee -a ipmisol-log)) if $bmwqemu::vars{IPMI_SOL_LOG};
```

**Review by Martchus — CHANGES_REQUESTED** [review]

**Comment by Martchus** [comment]
Thanks, I only have a minor nitpick.

**Comment by alice-suse** [comment]
Very good supplemental log for bare metal test. The log is well readable for text logs, while for text graphic, eg grub2 menu or autoyast installation process, the output is not easy to interpret directly. It will be good to find some tool to interpret that part, but it is a different story, the PR code is good to merge, IMO.

**Comment by alice-suse** [comment]
@mdoucha @czerw Maybe you are interested?

File: doc/backend_vars.asciidoc (hunk 1 above)
Comment (waynechen55) [reply doc/backend_vars.asciidoc]: Done

File: consoles/sshXtermIPMI.pm (hunk 2 above)
Comment (waynechen55) [reply consoles/sshXtermIPMI.pm]: Done

**Comment by Julie-CAO** [comment]
This enhancement is very helpful for test review! thank you.

**Comment by czerw** [comment]
Why? We have already....
Screenshot from 2024-06-20 06-06-20

**Comment by waynechen55** [comment]
> Why? We have already.... Screenshot from 2024-06-20 06-06-20
> 
> Introduced in 2021 #1881

I suggest remove ```HARDWARE_CONSOLE_LOG``` and keep ```IPMI_SOL_LOG``` if they are the same thing.

**Comment by waynechen55** [comment]
> Why? We have already.... Screenshot from 2024-06-20 06-06-20
> 
> Introduced in 2021 #1881

I tried running test with ```HARDWARE_CONSOLE_LOG=1``` which uploads similar log to ```uploaded log``` section instead of ```result files```. Please refer to this [test run with HARDWARE_CONSOLE_LOG=1 link1](http://10.67.18.153/tests/2857) [link2](http://10.100.228.134/tests/2857) [link3](http://10.100.204.18/tests/2857) [link4](http://10.149.193.170/tests/2857).

**Comment by czerw** [comment]
I can't access any VR link on http://10.67.18.153/ and I'm in company VPN as usual.

**Comment by waynechen55** [comment]
> I can't access any VR link on http://10.67.18.153/ and I'm in company VPN as usual.

Use link2. Try another link if you can not access either of them.

**Comment by czerw** [comment]
> Use link2. Try another link if you can not access either of them.

This one works for me http://10.100.228.134/tests/2857/logfile?filename=hardware-console-log.txt  and we can see `HARDWARE_CONSOLE_LOG` works for you.

**Comment by waynechen55** [comment]
> > Use link2. Try another link if you can not access either of them.
> 
> This one works for me http://10.100.228.134/tests/2857/logfile?filename=hardware-console-log.txt and we can see `HARDWARE_CONSOLE_LOG` works for you.

* First it is not in ```Result files``` section.
* Second I think it would be better to rename it to ```IPMI_SOL_LOG``` ???

**Comment by czerw** [comment]
> * First it is not in `Result files` section.

I don't see it as problem at all.

> * Second I think it would be better to rename it to `IPMI_SOL_LOG` ???

We need more generic name. It works also for other backends spvm, hmc (which don't use sol).

**Comment by waynechen55** [comment]
> > * First it is not in `Result files` section.
> 
> I don't see it as problem at all.
> 
> > * Second I think it would be better to rename it to `IPMI_SOL_LOG` ???
> 
> We need more generic name. It works also for other backends spvm, hmc (which don't use sol).

If this is the case, I think ```HARDWARE_CONSOLE_LOG``` can also be adopted by us.

**Review by okurz — CHANGES_REQUESTED** [review]
waiting for clarification on https://github.com/os-autoinst/os-autoinst/pull/2504#issuecomment-2180015212

**Comment by waynechen55** [comment]
> waiting for clarification on [#2504 (comment)](https://github.com/os-autoinst/os-autoinst/pull/2504#issuecomment-2180015212)

What to be clarified ?

**Comment by okurz** [comment]
> > waiting for clarification on [#2504 (comment)](https://github.com/os-autoinst/os-autoinst/pull/2504#issuecomment-2180015212)
> 
> What to be clarified ?

If you can just use the existing feature and close this PR unmerged, need to adapt the existing feature and update this PR or need this PR with the additional feature

**Comment by waynechen55** [comment]
> > > waiting for clarification on [#2504 (comment)](https://github.com/os-autoinst/os-autoinst/pull/2504#issuecomment-2180015212)
> > 
> > 
> > What to be clarified ?
> 
> If you can just use the existing feature and close this PR unmerged, need to adapt the existing feature and update this PR or need this PR with the additional feature

I think we can use existing feature.

**Comment by okurz** [comment]
as per https://github.com/os-autoinst/os-autoinst/pull/2504#issuecomment-2181864092

**Pull request closed without merging**